### ISSUE TITLE
Multiple MD files in one PO file

### ISSUE THREAD

**Issue by keunes** [mondeja/mdpo#169]
Hello!

For AntennaPod we're [looking into](https://forum.antennapod.org/t/french-translation-of-the-website/1454/7?u=keunes) options of translating our relatively new website. In my search I came across your tool.

I was wondering (didn't see anything about this in the docs): is it possible to list/input multiple MD files to end up with one PO file? That way I hope in Weblate/Pontoon we end up with one resource per website section, within the website project.

Many thanks anyway for your work on this!

**Comment by mondeja** [comment]
Sure. Given two files `bar.md` and `baz.md` in a directory named `foo/` whose contents are:

```markdown
# Bar
```

```markdown
# Baz
```

Executing `md2po --no-location 'foo/*.md'` will output:

```po
#
msgid ""
msgstr ""

msgid "Bar"
msgstr ""

msgid "Baz"
msgstr ""
```

Cheers!

**Comment by keunes** [comment]
I thought that was the holy grail. But actually, would it be possible to transform the updated PO file back into multiple MD files? We have one MD file per website page, and if from the single PO we can only get back a single MD file, we won't get far 0:)

**Comment by mondeja** [comment]
> actually, would it be possible to transform the updated PO file back into multiple MD files?

I'm not sure if I'm following you. From one PO file you can't get a MD file, you need another MD file as reference. PO files doesn't store the layout of the Markdown files, just the content of the blocks to translate.

> and if from the single PO we can only get back a single MD file, we won't get far 0:)

You can store multiple files translations in one PO file, see the next example. Given 2 files `foo.md` (with `# Foo\n`) and `bar.md` (with `# Bar\n`), executing:

```bash
$ md2po foo.md -po reference.po -sqm
$ md2po bar.md -po reference.po -sqm
```
The resulting `reference.po` file content will be:

```po
#
msgid ""
msgstr ""

#: foo.md:block 1 (header)
msgid "Foo"
msgstr ""

#: bar.md:block 1 (header)
msgid "Bar"
msgstr ""
```

Note that you can use this `reference.po` to translate both `foo.md` and `bar.md`, so you can use one file to translate all of your pages, like using a compendium in xgettext.

I hope this helps you, but I'm unsure about what you're asking and I can't analyze your environment to provide you the better translation strategy. If you provide a concrete situation I could give you more concrete indications.

**Comment by keunes** [comment]
Thanks a lot for all the quick replies, and sorry if I'm hard to follow. I'm mostly doing comms for the project so this technical stuff is, let's say, a nice challenge :)

So, to summarise the flow we're trying to achieve:
1. I update EN source file of our generic pages, located in our [`_i18n/en` folder](https://github.com/AntennaPod/antennapod.github.io/tree/master/_i18n/en)
2. when I accept a PR that updates the MD files, this triggers an update of a PO file in a _translations_ branch
    * we have PO files in a separate branch to keep a clean commit history in the _main_ branch
    * automation is done using GitHub actions (the [workflow file](https://github.com/keunes/AP-website-copy/blob/translations/.GitHub/workflows/en-updates-to-translation-repo.yml) I came up with based on what I saw elsewhere - no idea if this code makes sense)
    * in the GitHub action, we use your tool, e.g. `md2po '_i18n/en/*.md' --save _i18n/*/main-pages.po` (the source for this is still the _main_ branch - have to figure out if/how the output PO files can be sent to the _translations_ branch)
    * our translations branch now holds the EN PO file
3. as the English PO file in our translations branch gets updated, GitHub informs Weblate (via a webhook) that source files have been updated, and Weblate pulls the updated PO file
4. translators do their work
5. once in a while, in Weblate I press 'Push', and (new/updated) translated PO files get committed to our _translations_ branch, to `_i18n/*/main-pages.po`
6. when the _translations_ branch receives the new/updated PO files, this triggers an update of the MD files in the _main_ branch (and a rebuild of the Jekyll website)
  * automation done again using GitHub actions
  * using your tool, `po2md` specifically, in the GitHub action

Hope that wasn't too long, and explains what we're trying to achieve :)

So:
* For step 2, based on your suggestion, I could use `md2po about.md 404.md download.md --po-filepath main-pages.po -sqm`. What does `-sqm` refer to? I can't find it in the mdpo docs.
* If I understand correctly, in step 6 I need to define also again the md files in order to get translations back in the right ones. In the [tutorial in the docs](https://mdpo.readthedocs.io/en/master/tutorial.html) I see this: `po2md README.md --pofiles locale/es/LC_MESSAGES/readme.po --quiet  --save locale/es/LC_MESSAGES/README.md`
  * I'm unsure why there's two references to README.md. Is the first the 'positional argument' I read about in the docs? Is that the 'source' MD (English), while the second is the 'target' MD (Spanish)?
  * How would this work with multiple MD files? Something like `po2md about.md 404.md download.md --pofiles _i18n/*/mainpages.po --save _i18n/*/about.md _i18n/*/404.md _i18n/*/download.md`

Many thanks again!

**Comment by mondeja** [comment]
> What does `-sqm` refer to? I can't find it in the mdpo docs.

`-sqm` are the options `--save` (`-s`), `--quiet` (`-q`) and `--merge-pofiles` (`-m`) compounded together.

> I'm unsure why there's two references to README.md. Is the first the 'positional argument' I read about in the docs? Is that the 'source' MD (English), while the second is the 'target' MD (Spanish)?

That's correct: `po2md` reads the `README.md` file and creates from it a file in `locale/es/LC_MESSAGES/README.md` using the translations located in `locale/es/LC_MESSAGES/readme.po`.

> How would this work with multiple MD files? 

Sorry, I'm seeing now that the `po2md` help is a bit confusing, `po2md` don't accept multiple files at once. I usually don't execute a single magic command to create files for each language, instead I ussually create Bash loops for language codes and files, but from [v0.3.65](https://github.com/mondeja/mdpo/releases/tag/v0.3.65) there is a 4th command line program named `md2po2md` which could do something similar to what do you want in a single execution for all files and languages passed, but it is designed to recreate the same layout of files and directories of the input, so I suppose that it wouldn't serve you for what you want to achieve.

**Comment by keunes** [comment]
Thanks again for your kind help. Very much appreciated!

> `md2po2md` could do something similar to what do you want in a single execution for all files and languages passed, but it is designed to recreate the same layout of files and directories of the input, so I suppose that it wouldn't serve you for what you want to achieve.

What I described is just a process we came up with, so if there's something else that works I'm more than happy to do that.

Imagine we'd use `md2po2md`, I have these general questions:
* When in the process are translations provided? Is it a) use `md2po2md` b) update translation PO file c) run `md2po2md` again (to 'sync' new translations from the PO files back into the MD files)?
* What do you mean with 'designed to recreate the same layout of files and directories of the input'? Taking the [example from the docs](https://mdpo.readthedocs.io/en/master/tutorial.html#simple-readme-file-translation) (`md2po2md README.md faq/FAQ-1.md faq/FAQ-2.md -l es -l fr -o "locale/{lang}/strings.po"`), would
```
.
├── locale
├── README.md
└── faq
    ├── FAQ-1.md
    └── FAQ-2.md
```
turn into
```
.
├── locale
│   └── es
│   │   ├── faq
│   │   |  ├── FAQ-1.md
│   │   |  └── FAQ-2.md
│   │   ├── README.md
│   │   └── strings.po
│   └── fr
│   │   ├── faq
│   │   |  ├── FAQ-1.md
│   │   |  └── FAQ-2.md
│   │   ├── README.md
│   │   └── strings.po
├── README.md
└── FAQ.md
```
?
* Given the input file structure would be repeated exactly, could I use `md2po2md -l es -l fr -o "locale/{lang}/strings.po"` instead?

Sorry for all these questions. I would be happy to, once I understand how the tool works, submit a PR to expand the docs a bit for newbies like me.

**Comment by mondeja** [comment]
> When in the process are translations provided?

I don't know what it means "are translations provided". When executing the tool, right? The workflow is always the same. You extract translations into PO files with `md2po`, update your translations in the PO files and recreate a translated Markdown file with `po2md`. `md2po2md` is just a convenient wrapped for `md2po` and `po2md`.

> What do you mean with 'designed to recreate the same layout of files and directories of the input'? Taking the example from the docs...

Please, I've limited time to help, do a simple test locally first.

**Comment by keunes** [comment]
> The workflow is always the same. You extract translations into PO files with `md2po`, update your translations in the PO files and recreate a translated Markdown file with `po2md`.

Here I see and understand 3 distinct steps. As you write, translating is the secondand - it can take (e.g.) three weeks.

But if step 1 (md2po) and step 3 (po2md) are executed in a single command (in a few seconds using md2po2md) I have no time to update the created PO files in those very seconds - so when do I provide translations?

> Please, I've limited time to help, do a simple test locally first.

Yes, I understand. I tried but unfortunately I'm stuck on an old Ubuntu. After adding a PPA with a newer Python version I got an error installing mdpo.
@ByteHamster, it'd be great if you could have a look.

**Comment by mondeja** [comment]
> Here I see and understand 3 distinct steps. As you write, translating is the secondand - it can take (e.g.) three weeks.
>
>But if step 1 (md2po) and step 3 (po2md) are executed in a single command (in a few seconds using md2po2md) I have no time to update the created PO files in those very seconds - so when do update translations in the PO files?

If you've updated the translations, the `md2po`-part of `md2po2md` will keep them and will do nothing.

> I got an error installing mdpo

Maybe it's a bug. Do you've Python3.6 or grater installed? If so, could you paste the error traceback here, please?
